### PR TITLE
Enhancements, fix #98, #129

### LINKED ISSUE #98: Use Sort::getCriteria instead of Sort::getOrder
https://github.com/yiisoft/yii-cycle/blob/a67841b37cdc4ef2c89912bfcd9f86a0237abab5/src/Data/Reader/EntityReader.php#L208

### PR DESCRIPTION
**Pull request by rustamwin**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️/❌
| New feature?  | ✔️/❌
| Breaks BC?    | ✔️/❌
| Fixed issues  | #98, #129, #123

### CODE REVIEW

File: src/Data/Reader/EntityReader.php [hunk 1]
```diff
@@ -226,4 +226,19 @@ private function resetCountCache(): void
...
     }
+
+    private function normalizeCriteria(array $criteria): array
+    {
+        foreach ($criteria as $field => $direction) {
+            if (is_int($direction)) {
+                match ($direction) {
+                    SORT_DESC => $direction = 'DESC',
+                    default => $direction = 'ASC',
+                };
```
Comment (roxblnfk) [inline src/Data/Reader/EntityReader.php:237]: Can be simplified

File: src/Data/Reader/EntityReader.php [hunk 2]
```diff
@@ -226,4 +226,19 @@ private function resetCountCache(): void
...
         $this->countCache = new CachedCount($newQuery);
     }
+
+    private function normalizeCriteria(array $criteria): array
+    {
+        foreach ($criteria as $field => $direction) {
+            if (is_int($direction)) {
+                match ($direction) {
+                    SORT_DESC => $direction = 'DESC',
+                    default => $direction = 'ASC',
```
Comment (roxblnfk) [inline src/Data/Reader/EntityReader.php:236]: are you sure the `$direction` var can not contain other values like `"asc"` or `"desc"`?

File: src/Command/Migration/DownCommand.php [hunk 3]
```diff
@@ -18,21 +18,16 @@
 final class DownCommand extends BaseMigrationCommand
 {
     protected static $defaultName = 'migrate/down';
+    protected static $defaultDescription = 'Rolls back the last applied migration';
 
-    private EventDispatcherInterface $eventDispatcher;
+    private ?EventDispatcherInterface $eventDispatcher;
 
-    public function __construct(CycleDependencyProxy $promise, EventDispatcherInterface $eventDispatcher)
+    public function __construct(CycleDependencyProxy $promise, ?EventDispatcherInterface $eventDispatcher = null)
```
Comment (roxblnfk) [inline src/Command/Migration/DownCommand.php:25]: Could you tell me why eventDispatcher must be optional?

File: src/Data/Reader/EntityReader.php (hunk 2 above)
Comment (rustamwin) [reply src/Data/Reader/EntityReader.php:237]: Yes, checked for int in above.

File: src/Data/Reader/EntityReader.php (hunk 1 above)
Comment (rustamwin) [reply src/Data/Reader/EntityReader.php:238]: ?

File: src/Command/Migration/DownCommand.php (hunk 3 above)
Comment (rustamwin) [reply src/Command/Migration/DownCommand.php]: Reverted.

File: src/Data/Reader/EntityReader.php (hunk 1 above)
Comment (roxblnfk) [reply src/Data/Reader/EntityReader.php:238]: > ?

```php
                $direction = match ($direction) {
                    SORT_DESC => 'DESC',
                    default => 'ASC',
                };
```

Comment (rustamwin) [reply src/Data/Reader/EntityReader.php]: Done

File: src/Data/Reader/EntityReader.php [hunk 4]
```diff
@@ -226,4 +227,19 @@ private function resetCountCache(): void
         }
         $this->countCache = new CachedCount($newQuery);
     }
+
+    private function normalizeCriteria(array $criteria): array
```
Comment (roxblnfk) [inline src/Data/Reader/EntityReader.php:231]: Please rename this method to relate it with sorting. For example:
`normalizeSorting`,  `normalizeSortCriteria`, `getSortingValues`...

File: src/Data/Reader/EntityReader.php [hunk 5]
```diff
@@ -48,7 +48,7 @@ public function __construct(Select|SelectQuery $query)
             new Processor\LessThan(),
             new Processor\LessThanOrEqual(),
             new Processor\Like(),
-            // new Processor\Not()
+        // new Processor\Not()
```
Comment (roxblnfk) [inline src/Data/Reader/EntityReader.php:51]:
```suggestion
            // new Processor\Not()
```

File: src/Data/Reader/EntityReader.php [hunk 6]
```diff
@@ -184,14 +185,14 @@ private function setFilterProcessors(FilterProcessorInterface ...$filterProcesso
         $this->filterProcessors = array_merge($this->filterProcessors, $processors);
     }
 
-    private function buildQuery()
+    private function selectQuery(): SelectQuery|Select
```
Comment (roxblnfk) [inline src/Data/Reader/EntityReader.php:188]: This method builds a select query but not selects a query.
If you don't like the `buildQuery` method name then you can set it like `buildSelectQuery`. 
For the consistency! :)

**Comment by roxblnfk** [comment]
Looks ok

**Comment by samdark** [comment]
@rustamwin is it possible to fix CI?

**Comment by roxblnfk** [comment]
> @rustamwin is it possible to fix CI?

Not in this PR

**Comment by samdark** [comment]
👍

**Pull request merged**